### PR TITLE
perf: chat scrolling and markdown rendering

### PR DESCRIPTION
**Pull request by cogwheel0**
## Summary

Two rounds of performance work on the chat timeline and the streaming markdown pipeline, based on a full audit of both. No behavior changes intended aside from the deliberate ones listed below.

### Scrolling: stop shell rebuilds from cascading into every mounted row

- The timeline `SliverList`s used a fresh `SliverChildBuilderDelegate` per build, whose default `shouldRebuild` always returns true — every `ChatPage` `setState` (drag start, keyboard-inset frames, composer resize, pin transitions) rebuilt every mounted row and its markdown. They now use `_TimelineRowDelegate` with a real `shouldRebuild` keyed on rowBuilder/entries identity and `centerIndex`.
- `ChatPage` memoizes the transcript window, `ChatTimelineRenderModel`, and the `rowBuilder` closure by identity. The fresh window list allocated per build was also defeating the existing stable-layout cache's identity fast path.
- Full `MediaQuery.of` dependencies replaced with scoped `paddingOf`/`sizeOf` lookups.
- Streaming `cacheExtent` raised 120px → 600px: the small extent evicted rows on scroll-up mid-stream, and remounting a settled row runs a synchronous markdown compile, so it bought hitches, not savings.

### Markdown: cut O(whole-message) work per streamed flush

- `buildMarkdownDisplayParts` (per-block sub-document derivation + deep compares) is memoized by compiled-document identity; previously it ran on every widget build, ~10x/s during streaming.
- `identical()` fast paths in `CompiledMarkdownDocument`/`PreparedMarkdownText` equality; rope equality compares segments instead of materializing both sides.
- The chat structure signature caches per-message fragments in an identity-keyed `Expando` instead of rebuilding O(messages × versions) strings on every message-list emission.
- Incremental preparation and the reference-definition strip now gate on one shared line-anchored predicate (`hasLinkReferenceDefinitionLine`). A bare `]:` substring (code samples, mid-line citations) no longer forces full re-preparation of the whole message on every flush.
- The streaming split's unsafe-line detection is fence-aware and single-pass: definition-shaped lines or raw HTML inside code fences no longer disable incremental compilation for the entire message, and a raw HTML block now caps freezing at its start line instead of keeping the whole document mutable (HTML cannot affect earlier blocks). The per-fenced-block `RegExp` compile was replaced with the existing fence-close helper.
- LaTeX extraction skips its four full regex passes when content has no `$` or `\`; the assistant footer's error-heuristic content scans are memoized; profiler data-map allocations are gated to profiling builds; the code-block line split is memoized and >50k-char code renders plain (bounds `highlight.parse` cost for expanded still-streaming blocks).

### Deliberate behavior changes

- Content containing `]:` without an actual reference-definition line no longer gets the incidental 3+-newline collapse and trim from the strip branch (render-equivalent in markdown).
- Code blocks over 50k characters render without syntax highlighting in any language.
- Streaming and idle scroll cache extents are now both 600px.

## Testing

- `flutter analyze` clean (only the pre-existing `tool/pigeon_codegen` package-context errors).
- Full `flutter test`: the only failures are the 11 pre-existing ones on Flutter 3.47 (9 `direct_connections` `scrollUntilVisible` framework-behavior failures, 2 `release_notes` golden mismatches), verified identical on a clean `main` via `git stash`.
- New/updated tests: splitter coverage for fenced reference definitions, raw-HTML freeze capping, and the updated cache-extent expectation.

## Summary by CodeRabbit

* **New Features**
  * Added support for displaying valid follow-up suggestions received through chat updates.
  * Improved streaming Markdown handling for raw HTML, link references, details blocks, code blocks, and incremental updates.

* **Performance Improvements**
  * Improved chat timeline scrolling and rendering smoothness during streaming.
  * Reduced unnecessary processing for repeated messages, error indicators, LaTeX, and Markdown.
  * Improved handling of large code blocks and repeated content updates.

* **Bug Fixes**
  * Prevented stale, duplicate, or shorter streaming updates from truncating newer content.
  * Preserved buffered and structured content when streaming is cancelled or completed.
  * Improved HTML entity handling, whitespace preservation, and Markdown escaping.
  * Prevented semantic reasoning details from affecting content comparisons.
  * Kept final streaming status updates visible when appropriate.

---

### Long responses truncated at completion

The streamed buffer is never periodically folded into message state, so `/api/chat/completed` was built from a stale prefix of the response, and the server's echo of that payload was merged back unguarded — overwriting the full content with the prefix (worst on the HTTP/SSE transport, which reached completion without any terminal flush; the truncated value then persisted). Fixed by flushing the buffer before building the completed payload, and guarding the three unguarded overwrite paths (completed echo, replay-gap authoritative recovery, cumulative `chat:completion` content snapshots) so a strict prefix of already-streamed content is never adopted. Genuine outlet-filter rewrites (non-prefix content) still apply.

### Follow-up suggestions never arriving

The server emits `chat:message:follow_ups` strictly after `chat:completion {done:true}` (the follow-up LLM call runs post-done), and the per-stream socket subscription is disposed synchronously by the done event — so the streaming handler for follow-ups was unreachable in production. The long-lived passive conversation subscription is the surviving delivery path; it now applies the pushed payload directly to the target message instead of relying on a debounced full refetch that races the server's own persistence (the event is emitted before the upsert). Known limitation: temporary (`local:`) chats still can't receive follow-ups — the server doesn't persist them for unsaved chats and the passive subscription doesn't cover them.

New regression tests: completed-payload flush ordering and echo-truncation guard on the HTTP/SSE path (with a buffer-accurate test double), plus the follow-ups envelope parser.

<h3>Greptile Summary</h3>

Improves streamed Markdown rendering so reference definitions and nested details content are handled consistently while content arrives incrementally.

## T-Rex validation blocked
Flutter and Dart are unavailable in the environment, so the focused Markdown harness and the existing controller test could not execute.

<details><summary><h3>Confidence Score: 5/5</h3></summary>

No blocking failure remains.

No accepted blocking finding remains.
</details>

<details><summary><h3><a href="https://www.greptile.com/trex">T-Rex</a> T-Rex Logs</h3></summary>

**What T-Rex did**
- Attempted to run the focused streaming Markdown scenario test harness using Flutter in the repository, but Flutter CLI was not found and the command exited with code 127.
- Attempted to run the existing controller Flutter test for the focused scenario, but Flutter CLI was not found and the command exited with code 127.
- Attempted to fall back to a Dart test command for the streaming Markdown scenario, but the Dart tool was not found and the command exited with code 127.
- The before-capture baseline shows the base scanner’s loose close-token behavior and limited details-depth handling.
- Reviewed the Streaming Markdown PR diff and coverage capture to confirm test targets and mappings.

<a href="https://app.greptile.com/trex/runs/19259482/artifacts"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/ViewAllArtifactsDark.svg?v=4"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/ViewAllArtifacts.svg?v=4">View all artifacts</picture></a>

<sub><a href="https://www.greptile.com/trex">T-Rex</a> Ran code and verified through T-Rex</sub>
</details>

<sub>Reviews (20): Last reviewed commit: ["fix: count only complete details tags; k..."](https://github.com/cogwheel0/conduit/commit/7daaa2e972d61d591ae07dd7a1b7e9c65ecf33d5) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=53811184)</sub>

### CODE REVIEW

File: lib/shared/widgets/markdown/markdown_config.dart [hunk 1]
```diff
@@ -1246,27 +1246,45 @@ class _CodeBlockBody extends StatefulWidget {
...
-        widget.highlightLanguage == 'json' &&
-        (lines.length > _CodeBlockBody.largeJsonPlainPreviewLineThreshold ||
-            widget.code.length >
-                _CodeBlockBody.largeJsonPlainPreviewCharThreshold);
+        displayCode.length > _CodeBlockBody.largePlainRenderCharThreshold ||
+        (_isCollapsed &&
+            widget.highlightLanguage == 'json' &&
+            (lines.length > _CodeBlockBody.largeJsonPlainPreviewLineThreshold ||
+                widget.code.length >
+                    _CodeBlockBody.largeJsonPlainPreviewCharThreshold));
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply lib/shared/widgets/markdown/markdown_config.dart:1287]: Intentional: the cap bounds the synchronous highlight.parse cost of what is actually rendered, which is displayCode. A collapsed block renders only the 10-line preview, and highlighting that preview is cheap regardless of total block size; once expanded, displayCode is the full source and the cap applies. Checking widget.code.length would needlessly un-highlight collapsed previews of large blocks.

File: lib/shared/widgets/markdown/markdown_preprocessor.dart [hunk 2]
```diff
@@ -315,9 +315,22 @@ class ConduitMarkdownPreprocessor {
...
+  ///
+  /// This is the exact trigger for [_stripLinkReferenceDefinitions]'s global
+  /// rewrite (strip + newline collapse + trim), so the incremental streaming
+  /// preparation engine gates on the same predicate: whenever this is false,
+  /// normalization is guaranteed to have no non-local effects from reference
+  /// definitions. A bare `contains(']:')` was previously used for both, which
+  /// forced full re-preparation for any message merely mentioning `]:` (for
+  /// example inside code).
+  static bool hasLinkReferenceDefinitionLine(String input) =>
+      input.contains(']:') && _linkReferenceDefinition.hasMatch(input);
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply lib/shared/widgets/markdown/markdown_preprocessor.dart:335]: The predicate is deliberately not code-aware. When it is true, _stripLinkReferenceDefinitions still runs its global branch (newline collapse + trim) even if the masked strip removes nothing, so normalization does have non-local effects exactly when this predicate is true — the incremental-preparation gates and the preprocessor stay in precise parity by sharing it. A fence-aware predicate would have to track fences on raw, pre-normalized content (normalize fixes indented/inline fences only after the strip), which risks divergence in edge cases for a further optimization. The fence-aware handling lives in the compiled-split layer instead, where content is already normalized.

File: test/shared/widgets/markdown/markdown_document_controller_test.dart [hunk 3]
```diff
@@ -106,17 +106,59 @@ void main() {
...
+          'See [docs][d].\n\n<div>\nraw html\n</div>\n\n[d]: https://example.com';
+
+      final split = debugSplitStreamingPreparedContentForTesting(content);
+
+      expect(split['frozenPrefix'], isEmpty);
+      expect(split['mutableTail'], content);
+      expect(split['canIncrementallyCompile'], isFalse);
+      expect(split['fallbackReason'], 'referenceDefinitions');
+    },
+  );
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply test/shared/widgets/markdown/markdown_document_controller_test.dart:158]: Keeping expect here: every existing test in this file uses flutter_test's expect, and the new splitter tests follow the surrounding style. Converting only these three would make the file inconsistent; a wholesale migration to package:checks belongs in its own change.

**Comment by cogwheel0** [comment]
Re the outside-diff-range comment on `_firstStreamingUnsafeLine` (raw HTML state tracking): fixed in 8697467c, with a narrower approach than full CommonMark end-condition tracking. Once a raw HTML block has started, the scan stops trusting fence state and checks every subsequent line for a reference definition — so backtick lines inside raw HTML can no longer hide a later definition. Regression tests added for both scenarios (definition after raw HTML containing an odd backtick-line count, and definition-shaped content inside the raw block).

The second half of the suggestion — suppressing the definition fallback for definition-shaped lines *inside* pre/script/style/textarea — is intentionally not done: that needs real per-tag end-condition tracking to be sound, and the conservative whole-region fallback matches the pre-PR behavior for such content. Left as a follow-up if it ever shows up in practice.

File: lib/features/chat/providers/chat_providers.dart [hunk 4]
```diff
@@ -2470,6 +2530,27 @@ class ChatMessagesNotifier extends Notifier<List<ChatMessage>>
...
+    );
+    updateMessageById(parsed.messageId, (current) {
+      if (listEquals(current.followUps, parsed.followUps)) {
+        return current;
+      }
+      return current.copyWith(followUps: parsed.followUps);
+    });
+    return true;
+  }
+
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply lib/features/chat/providers/chat_providers.dart:2536]: Fixed in 7e6550f8: an unknown message id now returns false so the event falls through to the debounced refetch.

File: lib/features/chat/providers/chat_providers.dart [hunk 5]
```diff
@@ -82,6 +82,41 @@ part 'chat_composer_providers.dart';
...
+  ];
+  if (followUps.isEmpty) {
+    return null;
+  }
+  return (
+    messageId: messageId,
+    followUps: List<String>.unmodifiable(followUps),
+  );
+}
+
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply lib/features/chat/providers/chat_providers.dart]: Fixed in 7e6550f8: split into a private _parseFollowUpsSocketEvent used by production code, with debugParseFollowUpsSocketEventForTesting as a thin visibleForTesting wrapper, matching the file convention.

File: test/features/chat/providers/follow_ups_socket_event_test.dart [hunk 6]
```diff
@@ -0,0 +1,75 @@
+import 'package:conduit/features/chat/providers/chat_providers.dart';
+import 'package:flutter_test/flutter_test.dart';
+
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply test/features/chat/providers/follow_ups_socket_event_test.dart]: Done in 7e6550f8 — the new test file now uses package:checks.

File: lib/shared/widgets/markdown/markdown_document_controller.dart [hunk 7]
```diff
@@ -901,16 +903,37 @@ List<_StreamingPreparedLine> _splitStreamingPreparedLines(String content) {
...
+  // check this replaced.
+  for (final line in lines) {
     final candidate = _streamingBlockStarterCandidate(line.text);
     if (candidate == null) {
       continue;
     }
 
+    if (firstRawHtmlOffset != null) {
+      if (_streamingReferenceDefinitionPattern.hasMatch(candidate)) {
+        return (offset: line.start, isReferenceDefinition: true);
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply lib/shared/widgets/markdown/markdown_document_controller.dart]: Fixed in 7e6550f8: the pattern now consumes backslash-escaped characters inside the label (`^\s{0,3}\[(?:\\.|[^\]\n])+\]:`), with a regression test for an escaped-bracket label defined after a raw HTML block.

File: test/core/services/streaming_helper_transport_test.dart [hunk 8]
```diff
@@ -79,6 +79,64 @@ class _GatedChatCompletedApi extends ApiService {
...
+    required String model,
+    Map<String, dynamic>? modelItem,
+    String? sessionId,
+    List<String>? filterIds,
+    ApiAuthSnapshot? authSnapshot,
+  }) async {
+    capturedMessages = messages;
+    return {'messages': responseMessages ?? messages};
+  }
+}
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply test/core/services/streaming_helper_transport_test.dart:112]: Keeping the ApiService subclass double: this file's existing convention for exactly this seam is _GatedChatCompletedApi, an ApiService subclass overriding sendChatCompleted. A mocktail mock here would make the two doubles inconsistent within the same file; converting both belongs in a separate cleanup.

File: test/core/services/streaming_helper_transport_test.dart [hunk 9]
```diff
@@ -693,6 +751,104 @@ void main() {
...
+          await Future<void>.delayed(const Duration(milliseconds: 10));
+          return log;
+        }
+
+        final stalePrefix = await run('Hello');
+        check(stalePrefix.messages.last.content).equals('Hello world');
+
+        final rewrite = await run('REWRITTEN BY OUTLET');
+        check(rewrite.messages.last.content).equals('REWRITTEN BY OUTLET');
+      },
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply test/core/services/streaming_helper_transport_test.dart]: Declining the discriminator: the Open WebUI protocol carries no revision or marker distinguishing a stale echo from an intentional rewrite — the client cannot invent one unilaterally. With the payload now built from fully flushed state, the server echo is the complete content and genuine outlet rewrites (including shortening ones) round-trip as non-prefix changes in practice; an outlet filter that intentionally truncates the response to an exact prefix of itself is the one unsupported case, and losing streamed content by default was the worse trade.

File: test/core/services/streaming_helper_transport_test.dart [hunk 10]
```diff
@@ -3657,431 +3756,87 @@ void main() {
...
-
-        check(log.messages[1].content).equals('Old answer');
-        check(log.messages[3].id).equals('msg-2');
-        check(log.messages[3].content).equals('');
-        check(log.messages[3].followUps).isEmpty();
-        check(log.messages[3].statusHistory).isEmpty();
-        check(log.finishCount).equals(1);
-      },
-    );
+      }
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply test/core/services/streaming_helper_transport_test.dart:4051]: Out of scope for this PR — those six waits predate this change and none of the new tests use the delayed done-recovery path. Worth doing in a test-speed cleanup.

**Comment by cogwheel0** [comment]
Third streaming fix pushed (de2e3e90), addressing the remaining 'long responses still incomplete' report: the structured-output projector defers full re-renders geometrically (next re-render only at 2× the last projected length) and permanently disables its append fast path once the text contains a backtick — so on Open WebUI 0.11's output-snapshot streaming, visible content could trail the logical content by up to half the response. A plain delta arriving after a deferral then appended onto the short stale render and blocked the terminal repair, permanently dropping the deferred middle. Fixes: materialize the full projection before switching to plain appends (new syncProjectionToLatest), finalize the projector before building the /api/chat/completed payload, and fold the un-flushed buffer into state on stream cancel (conversation switch / deletion mid-stream). Regression test verified to fail without the fix.

**Comment by cogwheel0** [comment]
Pushed e07a6d56 completing the remaining streaming-integrity findings from the audit: content comparisons now strip rendered semantic <details> wrappers (local duration="0" vs server real-duration attributes made every raw prefix/length guard a no-op on reasoning turns) across applyServerContent, snapshot preservation, the completed-echo/replay-gap/cumulative-snapshot guards; Hermes keeps the streamed text when a terminal output is a strict prefix of it; the local turn-echo payload now carries output/sources/usage/followUps/files/embeds/statusHistory/error so a sync-outbox push can no longer wipe those fields from the server copy; and a stale settled markdown refresh can't strand the loading-skeleton flag.

File: lib/shared/widgets/markdown/markdown_document_controller.dart [hunk 11]
```diff
@@ -41,8 +42,10 @@ final RegExp _streamingRawHtmlBlockPattern = RegExp(
   caseSensitive: false,
   multiLine: true,
 );
+// Labels may contain backslash-escaped brackets ("[a\]b]:"), so escaped
+// pairs are consumed before the closing bracket.
 final RegExp _streamingReferenceDefinitionPattern = RegExp(
-  r'^\s{0,3}\[[^\]\n]+\]:',
+  r'^\s{0,3}\[(?:\\.|[^\]\n])+\]:',
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply lib/shared/widgets/markdown/markdown_document_controller.dart]: Fixed in bd027f11: the label class now excludes the backslash (`(?:\\.|[^\]\\\n])+`) so the branches cannot overlap; escaped labels like `[a\]b]:` remain covered by the existing regression test.

File: test/shared/widgets/markdown/markdown_document_controller_test.dart [hunk 12]
```diff
@@ -439,68 +506,65 @@ Body
...
+    addTearDown(compiler.dispose);
+    final engine = StreamingMarkdownPreparationEngine();
+    var prepared = const PreparedMarkdownText.empty();
+    CompiledMarkdownDocument? latestDocument;
+    final controller = MarkdownDocumentController(
+      readCompiler: () => compiler,
+      isWidgetTest: () => false,
+      onStateChanged: (document) => latestDocument = document,
+    );
+    addTearDown(controller.dispose);
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply test/shared/widgets/markdown/markdown_document_controller_test.dart:662]: These controller setups predate this PR — main has the same fourteen MarkdownDocumentController constructions; the dart format rewrap in this branch made them appear added. Extracting a harness is a reasonable cleanup for a dedicated test-refactor change, not this PR.

File: lib/shared/widgets/markdown/markdown_document_controller.dart [hunk 13]
```diff
@@ -927,12 +952,18 @@ bool _containsStreamingRawHtmlBlock(String content) {
       continue;
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply lib/shared/widgets/markdown/markdown_document_controller.dart:1015]: Fixed in bd027f11: the unsafe-line scan now tracks <details> bodies opaquely by open/close depth, mirroring the block scanner, so an unmatched backtick line inside a details body no longer opens a phantom outer fence. Regression tests cover the two-update scenario (definition after a details body with an unmatched fence) and that a closed details block still freezes normally.

File: lib/features/chat/providers/chat_providers.dart [hunk 14]
```diff
@@ -6180,6 +6310,24 @@ class ChatMessagesNotifier extends Notifier<List<ChatMessage>>
...
+          'sources': message.sources
+              .map((source) => source.toJson())
+              .toList(growable: false),
+        if (message.statusHistory.isNotEmpty)
+          'statusHistory': message.statusHistory
+              .map((status) => status.toJson())
+              .toList(growable: false),
+        if (message.followUps.isNotEmpty)
+          'followUps': List<String>.from(message.followUps),
+        if (message.error != null) 'error': message.error!.toJson(),
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply lib/features/chat/providers/chat_providers.dart]: Fixed in bd027f11: codeExecutions is now persisted in the echo payload via execution.toJson(). versions deliberately left out per the sibling-row derivation.

File: lib/core/services/streaming_helper.dart [hunk 15]
```diff
@@ -793,14 +794,19 @@ ActiveChatStream attachUnifiedChunkedStreaming({
...
+      final semanticDetailsPattern = RegExp(
+        r'''<details\b(?=[^>]*\btype\s*=\s*["'](?:reasoning|tool_calls|code_interpreter|openai_builtin_tool)["'])[\s\S]*?</details>\s*''',
+      );
+      plain = plain.replaceAll(semanticDetailsPattern, '').trim();
     }
-    final semanticDetailsPattern = RegExp(
-      r'''<details\b(?=[^>]*\btype\s*=\s*["'](?:reasoning|tool_calls|code_interpreter|openai_builtin_tool)["'])[\s\S]*?</details>\s*''',
-    );
-    return content.replaceAll(semanticDetailsPattern, '').trim();
+    return plain.contains('&') ? plainContentUnescape.convert(plain) : plain;
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply lib/core/services/streaming_helper.dart]: Fixed in 77173396: the strip now consumes only the wrapper's own trailing newline (no \s*, no trim), so answer whitespace such as leading indented-code indentation survives.

File: lib/core/services/openwebui_stream_parser.dart [hunk 16]
```diff
@@ -185,8 +185,21 @@ Iterable<OpenWebUIStreamUpdate> parseOpenWebUIParsedPayload(
...
+  // delta's text (and its reasoning as an output item), so applying both
+  // duplicates content. Mute only when the snapshot parses into renderable
+  // blocks (matching the socket path's predicate): an output whose items all
+  // parse away would otherwise mute the delta AND render nothing.
+  final output = parsed['output'];
+  final outputUpdate = output is List && output.isNotEmpty
+      ? OpenWebUIOutputUpdate(output)
+      : null;
+  final hasOutputSnapshot =
+      outputUpdate != null && outputUpdate.blocks.isNotEmpty;
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply lib/core/services/openwebui_stream_parser.dart:199]: Added in 77173396: regression test asserting a whitespace-only (non-renderable) output snapshot does not mute the same-frame delta and yields an empty-blocks update that downstream handlers already ignore (applyParsedOpenWebUIUpdate requires non-empty normalized output, and replaceVisibleAssistantStructuredOutput bails on empty blocks).

File: lib/core/services/structured_output_renderer.dart [hunk 17]
```diff
@@ -293,7 +323,16 @@ final class StructuredOutputStreamingProjector {
...
+    // With appends available, geometric backoff is safe — appends carry the
+    // tail between full renders. Once a backtick/tilde disables appends for
+    // the rest of the turn, doubling would leave the visible tail up to 50%
+    // behind until completion; fall back to a bounded additive step instead.
+    _nextFullProjectionLength = logicalLength == 0
+        ? 1
+        : _appendIsPlain
+        ? logicalLength * 2
+        : logicalLength +
+              ((logicalLength >> 3) > 64 ? (logicalLength >> 3) : 64);
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply lib/core/services/structured_output_renderer.dart:340]: The named test still holds its geometric bound: reasoning-only streams have no plain content with code delimiters, so _appendIsPlain stays true and doubling applies — it passes unchanged. The additive schedule only engages for code-bearing plain content, and 77173396 adds a dedicated test documenting that bound (projection count > 14 but <= 64 for a 4096-char backtick-bearing stream, exact terminal render). The denser schedule there is deliberate: bounded staleness for streams whose append path is disabled.

File: lib/shared/widgets/markdown/markdown_document_controller.dart [hunk 18]
```diff
@@ -901,16 +908,57 @@ List<_StreamingPreparedLine> _splitStreamingPreparedLines(String content) {
...
+        return (offset: line.start, isReferenceDefinition: true);
+      }
+      continue;
+    }
+
+    if (detailsDepth > 0) {
+      detailsDepth += _streamingDetailsOpenPattern.allMatches(candidate).length;
+      detailsDepth -= _streamingDetailsClosePattern
+          .allMatches(candidate)
+          .length;
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply lib/shared/widgets/markdown/markdown_document_controller.dart:969]: Fixed in c24c8184: both streaming details trackers (the unsafe-line scan and the preparation checkpoint scanner, which had the same loose pattern) now match the parser's exact close token `</details>`. Regression test covers the lookalike-close scenario and was verified to fail against the loose pattern.

**Comment by cogwheel0** [comment]
@coderabbitai review

File: lib/shared/widgets/markdown/markdown_document_controller.dart [hunk 19]
```diff
@@ -901,16 +913,57 @@ List<_StreamingPreparedLine> _splitStreamingPreparedLines(String content) {
...
+    if (firstRawHtmlOffset != null) {
+      if (_streamingReferenceDefinitionPattern.hasMatch(candidate)) {
+        return (offset: line.start, isReferenceDefinition: true);
+      }
+      continue;
+    }
+
+    if (detailsDepth > 0) {
+      detailsDepth += _streamingDetailsOpenPattern.allMatches(candidate).length;
+      detailsDepth -= _streamingDetailsClosePattern
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply lib/shared/widgets/markdown/markdown_document_controller.dart:967]: Fixed in 1bfa3776 — depth counting now uses the parser's complete-tag pattern (`<details(?:\s+[^>]*)?>`, non-anchored) in both the freeze scanner and the unsafe-line pass, with a carve-out so a line-leading partial `<details` still opens scanner details mode. Block entry detection stays line-anchored to match the parser. Regression test added: an inline nested open followed by a body backtick no longer hides a later reference definition (verified failing on the old counting).

File: test/core/services/openwebui_stream_parser_test.dart [hunk 20]
```diff
@@ -1034,9 +1071,31 @@ void main() {
...
+        projector.project([StructuredOutputTextBlock(text: text.toString())]);
+      }
+
+      check(projector.fullProjectionCount).isGreaterThan(14);
+      check(projector.fullProjectionCount).isLessOrEqual(64);
+
+      final completion = projector.finish();
+      check(completion).isA<StructuredOutputStreamingReplace>();
+      check(completion!.content).endsWith('x');
+      check(completion.content.length).isGreaterOrEqual(chunkCount);
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply test/core/services/openwebui_stream_parser_test.dart]: Fixed in 1bfa3776 — the test now asserts exact equality with the full expected content instead of suffix/length checks.

**Comment by cogwheel0** [comment]
Re the outside-diff-range comment on `markdown_document_controller.dart` (reference definitions inside `<details>` forcing the whole-document fallback): fixed in 8797b096. Verified the premise first — the details parser lifts the body into a `body_markdown` attribute compiled as its own document, so definition-shaped lines inside a body are not document-scoped and cannot couple frozen and mutable segments. The check is removed from the depth-tracking branch, with a regression test covering `[d]: ...` inside a closed details block splitting normally.

8797b096 also lands a broader audit pass: server-shape `sources`/`code_executions`/`files` in the local-echo and direct/Hermes persisted payloads (the OWUI web client reads citation shape and snake_case), an observe-only guard in `syncProjectionToLatest`, verbatim user-message copies, the plain-accumulator invariant in `replaceVisibleAssistantContent`'s fallback, and shared `semantic_details.dart`/`openwebui_message_payload.dart` helpers replacing five divergent regex copies and five hand-rolled stale-prefix guards. a28040aa is a mechanical repo-wide `dart format` migration commit (no semantic changes).

**Comment by cogwheel0** [comment]
@coderabbitai review

**Comment by cogwheel0** [comment]
@greptile-apps review

**Comment by cogwheel0** [comment]
@greptileai review

File: lib/shared/widgets/markdown/markdown_document_controller.dart [hunk 21]
```diff
@@ -901,16 +933,57 @@ List<_StreamingPreparedLine> _splitStreamingPreparedLines(String content) {
...
 
+    if (firstRawHtmlOffset != null) {
+      if (_streamingReferenceDefinitionPattern.hasMatch(candidate)) {
+        return (offset: line.start, isReferenceDefinition: true);
+      }
+      continue;
+    }
+
+    if (detailsDepth > 0) {
+      detailsDepth += _streamingDetailsOpenCount(candidate);
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply lib/shared/widgets/markdown/markdown_document_controller.dart]: Fixed in 58d878da — depth counting in both the freeze scanner and the unsafe-line pass now runs on the raw line text (matching the parser, which counts tags regardless of indentation inside a block) instead of the dedented block-starter candidate. The partial-open carve-out dedents before its anchored check so a <=3-space-indented incomplete tag still holds the block open. Regression test added with an indented nested open + body backtick + trailing definition, verified failing on the previous counting.

**Comment by cogwheel0** [comment]
@greptile-apps review

File: lib/shared/widgets/markdown/markdown_document_controller.dart [hunk 22]
```diff
@@ -29,8 +30,37 @@ final RegExp _streamingDetailsOpenPattern = RegExp(
...
+  final count = _streamingDetailsInlineOpenPattern.allMatches(line).length;
+  if (count == 0) {
+    // A block-starting `<details` whose tag is still incomplete (no `>` yet)
+    // must still open scanner details mode: the parser will once the tag
+    // completes, and closing here would freeze an unterminated block. The
+    // line is dedented first so a <=3-space-indented partial open counts.
+    final candidate = _streamingBlockStarterCandidate(line);
+    if (candidate != null && _streamingDetailsOpenPattern.hasMatch(candidate)) {
+      return 1;
+    }
```
[the remark this answers is not in the document]
Comment (cogwheel0) [reply lib/shared/widgets/markdown/markdown_document_controller.dart]: Fixed in 7daaa2e9 — depth counting is now parser-exact (complete `<details ...>` tags only, no partial-tag credit), so a literal `<details` in a body that never completes no longer leaves the scanner inside a block the parser closed. The case the partial credit existed for — an unterminated streamed ENTRY tag — is handled structurally instead: an entry line whose tag has no closing `>` keeps the tail mutable. Regression tests cover both directions (stale depth past the close, and the incomplete entry staying mutable), each verified failing on the previous counting.

**Comment by cogwheel0** [comment]
@greptile-apps review
